### DISCUSSION TITLE
MicroPython library for text to speech conversion

### DISCUSSION THREAD

**Discussion by MehtaDhairya** [micropython/micropython#14378] [Libraries & Drivers]
I am trying to write a MicroPython code to convert text to speech using a esp32 board, LM386 circuit and a speaker, but i can't find a library for it and i also can't use API because most of them are paid. Is there any library i can use in MicroPython ?

**Comment by scruss** [comment]
Any particular language? Are you expecting the quality of online TTS systems? Hardware is typically not so good.

If English (or Mandarin) TTS with some limitations works for you, I wrote a crude library to drive a YuTone VoiceTX SYN6988 board: [YuTone VoiceTX SYN6988](https://github.com/scruss/micropython-SYN6988). These are fairly cheap, and include an amplifier. It only has one voice.

Much more expensive (and possibly discontinued) is the Parallax EMIC2 board. It runs an embedded version of  DECTalk 5. It uses a simple serial interface. It sounds a little robotic, but the advantage that DECTalk has for real TTX applications is that it can still be intelligible at very high words-per-minute values. 

(I seem to remember that the micro:bit version of MicroPython included the (ancient, terrible) "SAM" TTS system. It really doesn't sound like words at all.)

**Comment by deshipu** [comment]
MicroPython on the Micro:bit has the `speech` module that is using an old C64 library called SAM. I bet it would be possible to port it to other versions of MicroPython.

**Reply by scruss** [reply]
Have you heard it, though? It's unintelligible. Wouldn't be worth the effort.

**Reply by knghtbrd** [reply]
This is true if you're feeding it text. If you're feeding it phonemes it's … better. It's not amazing, but it's better. dectalk/dectalk exists and was even ported to the RPi, so for my own purposes the idea of grabbing one of my old circa 2012 boards that aren't useful for much else now feeding speech commands over the UART is kind of a serious consideration. But I also know that synthesizers I prefer more (such as the Keynote Gold) were originally based around the 80186. NEC cloned the 80186 as the V20 and V30 to upgrade IBM XT-class machines, and the Keynote Gold was miniaturized into a board that fit into the modem slot of the Toshiba T1000LE … so I'm pretty sure that decent speech synthesis well beyond SAM is possible on any ARM-based microcontroller, possibly using added flash to hold additional ROM,, but not much more than that.

The hunt continues.

**Reply by scruss** [reply]
The trouble with phonemes is that you need to pre-calculate them (or port Silas S. Brown's [lexconvert](https://github.com/ssb22/lexconvert) to MicroPython). So you're stuck with a fixed vocabulary.

I'm well aware of DECtalk. There is even a sort-of working demo MicroDECtalk that runs on a Raspberry Pi Pico. It takes up the whole firmware, though, and is currently hardwired to say one thing in terrible quality via PWM. Like SAM, the ports of DECtalk are abandonware, so they are difficult to integrate into a product. I suspect anything to do with Keynote Gold would have similar problems.

MikroE's [Text To Speech Click](https://www.mikroe.com/text-to-speech-click) gives you DECtalk 5 in hardware via the Epson-licensed chip. It's not cheap. Also also don't know if it suffered from the same amplifier-related sound quality problems that the Parallax board that used the same chip had.

If you can dedicate an entire ESP32 to speech generation, it looks like [pschatzmann/arduino-espeak-ng](https://github.com/pschatzmann/arduino-espeak-ng) could be useful. It is clearly open source, unlike many of the other options

**Comment by bidrohini** [comment]
Yes. You can use MicroPython TTS module. 
https://www.electromaker.io/project/view/micropython-tts-weather-broadcast
https://microbit-micropython.readthedocs.io/en/latest/tutorials/speech.html
If you are using a pico or ESP module, you can make a compact talking robot like this:
https://www.pcbway.com/project/shareproject/Batuni_The_Talkative_Messenger_Robot_b23d7996.html

**Reply by scruss** [reply]
The first of these uses the Google TTS API, which is no longer available.

The second one uses micro:bit speech, which is only on that platform and sounds dreadful.

The third one is an Arduino project, and unrelated to MicroPython.

**Comment by Josverl** [comment]
Not free, but you don't have to pay for the first $200.
Which will get you a long way.

https://azure.microsoft.com/en-us/products/ai-services/text-to-speech

**Reply by scruss** [reply]
MicroPython support? Looks like it might return a WAV or MP3, which a MicroPython board will struggle to play

**Reply by jonnor** [reply]
Nothing conceptually against playing back a WAV on a MicroPython board? There are examples in https://github.com/miketeachman/micropython-i2s-examples

**Comment by jacklinquan** [comment]
https://github.com/jacklinquan/micropython-samtts

A Python port of Software Automatic Mouth Text-To-Speech program.

SAM is the Text-To-Speech (TTS) software SAM (Software Automatic Mouth) for the Commodore C64 published in the year 1982 by Don't Ask Software (now SoftVoice, Inc.).

This project is an unofficial Python port of SAM. It is translated by hand from the adaption to C by [Stefan Macke](https://github.com/s-macke/SAM) and the refactorings by [Vidar Hokstad](https://github.com/vidarh/SAM).

As @scruss mentioned, its sound quality is not that good. But I believe it has its use. It adds text-to-speech functionality to very constrained hardware.

**Reply by scruss** [reply]
Thank you for unleashing this code on an unsuspecting world!

There's an ongoing project to port microDECTalk to the Raspberry Pi Pico. It has a similar "Abandonware / don't ask" licence. It's a long way off being usable yet
